### PR TITLE
add negative test case for test_convolve.py and test_astype.py

### PR DESCRIPTION
**Pull request by xialu00**
add negative test case for test_convolve.py

### CODE REVIEW

File: tests/integration/test_astype.py [hunk 1]
```diff
@@ -39,8 +61,30 @@ def test_basic(src_dtype, dst_dtype):
...
+    in_num = num.array(in_np)
+
+    out_np = in_np.astype(dst_dtype, order=order)
+    out_num = in_num.astype(dst_dtype, order=order)
+
+    assert np.array_equal(out_num, out_np)
+
+
+@pytest.mark.parametrize("src_dtype", ALL_BUT_COMPLEX[1:5], ids=to_dtype)
+@pytest.mark.parametrize("dst_dtype", ALL_BUT_COMPLEX[5:8], ids=to_dtype)
```
Comment (bryevdv) [inline tests/integration/test_astype.py:78]: This slicing seems fragile, if anyone happens to change the contents of `ALL_BUT_COMPLEX` without realizing it is used like this, this test could fail in subtle ways. I would make a new dedicated named lists, or just past list literals of the types here.

File: tests/integration/test_astype.py [hunk 2]
```diff
@@ -39,8 +61,30 @@ def test_basic(src_dtype, dst_dtype):
...
+@pytest.mark.parametrize("cast", CASTING, ids=str)
+def test_casting_negative(src_dtype, dst_dtype, cast):
+    in_num = num.array(TEST_VECTOR, dtype=src_dtype)
+    with pytest.raises(TypeError):
+        in_num.astype(dst_dtype, casting=cast)
+
+
+@pytest.mark.parametrize("src_dtype", ("F", "D"), ids=to_dtype)
+@pytest.mark.parametrize("dst_dtype", ALL_TYPES[1:], ids=to_dtype)
```
Comment (bryevdv) [inline tests/integration/test_astype.py:87]: same comment re: slicing being fragile

File: tests/integration/test_astype.py [hunk 3]
```diff
@@ -54,6 +98,24 @@ def test_complex(src_dtype, dst_dtype):
...
+    ]
+    in_np = np.array(complex_input, dtype=src_dtype)
+    in_num = num.array(in_np)
+
+    out_np = in_np.astype(to_dtype("?"))
+    out_num = in_num.astype(to_dtype("?"))
+    """
+    Numpy and cuNumeric have different performance.
+    For complex data 0.+1.j, Numpy set as True, cuNumeric set as False.
+    """
```
Comment (bryevdv) [inline tests/integration/test_astype.py:115]: why a string and not a comment?

File: tests/integration/test_convolve.py [hunk 4]
```diff
@@ -25,6 +25,29 @@
...
+def test_none():
+    # Numpy raises:
+    # TypeError: unsupported operand type(s) for *: 'NoneType' and 'NoneType'
+    with pytest.raises(AttributeError):
+        num.convolve(None, None, mode="same")
+
+
+def test_empty():
+    msg = r"empty"
+    with pytest.raises(ValueError, match=msg):
```
Comment (bryevdv) [inline tests/integration/test_convolve.py:38]: @manopapad do you think matching partial exception messages (optionally) is OK? Or do we strictly only want to check type going forward?

Comment (manopapad) [reply tests/integration/test_convolve.py:38]: This level of detail makes sense to me. You'd hope an error message about an empty array would contain the word "empty".

File: tests/integration/test_astype.py (hunk 3 above)
Comment (xialu00) [reply tests/integration/test_astype.py:115]: Fix in the form of comments

File: tests/integration/test_astype.py (hunk 2 above)
Comment (xialu00) [reply tests/integration/test_astype.py]: Redefined this variable

File: tests/integration/test_astype.py (hunk 1 above)
Comment (xialu00) [reply tests/integration/test_astype.py]: Redefined this variable

**Review by bryevdv — APPROVED** [review]

**Pull request merged**